cloud-controller-manager/do/lb_service_admission_handler.go

```diff
@@ -0,0 +1,166 @@
+	}
+
+	if svc.Spec.Type != corev1.ServiceTypeLoadBalancer {
+		return admission.Allowed("allowing service as it is not a load balancer")
+	}
+
+	if svc.DeletionTimestamp != nil {
+		return admission.Allowed("allowing service as it is being deleted")
+	}
+
+	lbReq, err := h.buildLoadBalancerRequest(&svc)
+	if err != nil {
+		return admission.Denied(fmt.Sprintf("building DO API request: %s", err))
+	}
+
+	var resp admission.Response
+	switch {
+	case req.OldObject.Raw != nil:
+		resp = h.validateUpdate(ctx, svc, req, lbReq)
+	default:
+		resp = h.validateCreate(ctx, svc, req, lbReq)
+	}
+
+	return resp
+}
+
+func (h *LBServiceAdmissionHandler) validateUpdate(ctx context.Context, svc corev1.Service, req admission.Request, lbReq *godo.LoadBalancerRequest) admission.Response {
+	var oldSvc corev1.Service
+	if err := h.decoder.DecodeRaw(req.OldObject, &oldSvc); err != nil {
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("decoding old object: %w", err))
+	}
+
+	lbID := oldSvc.Annotations[annDOLoadBalancerID]
+	if lbID == "" ||
+		len(oldSvc.Status.LoadBalancer.Ingress) == 0 ||
+		(len(oldSvc.Status.LoadBalancer.Ingress) > 0 && oldSvc.Status.LoadBalancer.Ingress[0].IP == "") {
+		// If service doesn't have an Ingress configured yet, it isn't fully configured and could have failed
+		// previously. We validate creation in that case.
+		return h.validateCreate(ctx, svc, req, lbReq)
+	}
+
+	_, resp, err := h.godoClient.LoadBalancers.Update(ctx, lbID, lbReq)
+	return h.mapGodoRespToAdmissionResp(resp, err)
+}
+
+func (h *LBServiceAdmissionHandler) validateCreate(ctx context.Context, svc corev1.Service, req admission.Request, lbReq *godo.LoadBalancerRequest) admission.Response {
+	_, resp, err := h.godoClient.LoadBalancers.Create(ctx, lbReq)
+	return h.mapGodoRespToAdmissionResp(resp, err)
+}
+
+func (h *LBServiceAdmissionHandler) buildLoadBalancerRequest(svc *corev1.Service) (*godo.LoadBalancerRequest, error) {
+	lbReq, err := buildLoadBalancerRequest(svc)
+	if err != nil {
+		return nil, err
+	}
+	lbReq.ValidateOnly = true
+	lbReq.Region = h.region
+	lbReq.VPCUUID = h.vpcID
+	lbReq.Tags = []string{buildK8sTag(h.clusterID)}
+	return lbReq, nil
+}
+
+func (h *LBServiceAdmissionHandler) mapGodoRespToAdmissionResp(resp *godo.Response, err error) admission.Response {
+	switch {
+	case err == nil:
+		return admission.Allowed("valid load balancer definition")
+	case resp == nil: // This case should never happen, asserting to prevent nil pointer exceptions.
+		return admission.Errored(http.StatusInternalServerError, errors.New("expected a DO API response"))
+	case resp.StatusCode < 400:
+		return admission.Allowed("valid load balancer definition")
+	case resp.StatusCode < 500:
+		return admission.Denied(fmt.Sprintf("invalid load balancer definition: %s", err))
+	default:
+		return admission.Allowed(fmt.Sprintf("received unexpected status code (%d) from DO API, allowing to prevent blocking: %s", resp.StatusCode, err))
```

Comment:
As discussed via DM, shouldn't it be okay here to return `admission.Errored` or `admission.Denied` and still get the request allowed if the failure policy is set to `Ignore` (and likewise get it rejected should users choose to go with `Fail`)?

Reply:
Unfortunately, returning `admission.Errored` or `admission.Denied` isn't ignored when we set the failure policy to `Ignore`. I dug through the [webhook admission code](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go#L170) to find the true meaning of the failure policy. It will ignore client errors such as: the webhook isn't responsive, etc.

For those reasons, I think the best solution is to:
- Configure the failure policy to `Ignore` to catch any potential failures when getting an answer from the webhook
- Have the webhook `Allow` the requests when faced with a 5xx (transient failure)
- Have the webhook `Deny` the requests when faced with a 4xx (non-transient failures)
- Have the webhook `Error` for unexpected errors within the whole logic.